### ISSUE TITLE
Grammar of buildfile for VSCode/ide syntax highlighting?

### ISSUE THREAD

**Issue by alexchandel** [build2/build2#109]
What's the grammar of the `buildfile`?  Creating a VSCode extension for that & the `manifest` would greatly help adoption, as it's the most popular editor.

**Comment by boris-kolpackov** [comment]
> Creating a VSCode extension for that & the `manifest` would greatly help adoption

Agree. @Klaim is also interested in that.

Let's first get `manifest` out of the way since it's pretty simple and is specified in [Manifest Format](https://build2.org/bpkg/doc/build2-package-manager-manual.xhtml#manifest-format).

Things are more complicated with `buildfile`. There is a high-level description in [Buildfile Language](https://build2.org/build2/doc/build2-build-system-manual.xhtml#intro-lang) but there is no formal grammar yet. Also, creating the precise grammar won't be easy because Buildfile is context-dependent (i.e., the set of characters that are recognized as special changes depending on the language construct). While this makes for a language that is pleasant for humans to write (IMO, anyways), it also makes it more difficult for machines to parse. We've tried to provide a precise grammar for Testscript, which is also context-dependent, and it got quite hairy (see [Lexical Structure](https://build2.org/build2/doc/build2-testscript-manual.xhtml#lexical) and [Syntax and Semantics](https://build2.org/build2/doc/build2-testscript-manual.xhtml#syntax)).

I think we will still produce the precise grammar at some point but, due to the complexity, I wonder how useful it will be to someone who is writing an IDE plugin. So perhaps the way forward would be to create a high-level/approximate grammar that can be used to recognize most of the constructs without implementing full context-sensitive parsing. What do you think?

BTW, there could also be something useful in [`build2-sublime-text`](https://github.com/build2/build2-sublime-text) (I haven't looked myself).

**Comment by Klaim** [comment]
When using VSCode today, I set it to use the MakeFile syntax highlighting on `buildfile`,  and `.build` files, it's already useful and can help figuring out some of the code faster. @alexchandel you might want to use this until a vscode extension is published.

About what would be necessary, it depends on the level of support to the language. My understanding is that there is:
 - syntax highlighing - depending on the language it can be easy or hard, but it can always be aproximated, like when using the MakeFile highlighing for `buildfile`;
 - auto-completion: this one imply to understand the language and the context, so it would need to interract with build2 in some ways to gather informations and/or to know the language constructs that are always available (like functions)
 - error highlighting: this requires a solid definition of the language and at least some of the semantic of the elements being manipulated by the syntax (it needs to know that you can't put a number as a pre-requisite, for example). This is also contextual.
 
There are other features related to syntax possible for an extension (like snippets) but that's for far in the future IMO.

**Comment by lyrahgames** [comment]
> BTW, there could also be something useful in [`build2-sublime-text`](https://github.com/build2/build2-sublime-text) (I haven't looked myself).

The Sublime Text syntax highlighting is only an approximation of the actual syntax.
I have only tried to map the most important high-level abstractions and
added support for toggling comments to make it more intuitive to work with buildfiles.
It is still work in progress.
I have attached a screenshot if you would like to take a look.
Currently, not every feature of the build system is supported or correctly highlighted.
Especially, ad-hoc recipes and a few other keywords are difficult to implement
because their documentation is still missing, not detailed enough, or because of the lack of time.

@boris-kolpackov:  Yes, I think it should be possible to completely
approximate syntax highlighting without precise grammar.
I agree with @Klaim with respect to auto-completion and error highlighting.
Furthermore, I believe it would be a good idea to provide
a simple auto-formatting tool for build2 like clang-format for C++.
In my opinion, such a tool maybe should be founded on the parser library of buildfiles.

![build2_syntax_highlight_example](https://user-images.githubusercontent.com/6022957/116168537-ec3b1400-a702-11eb-91ca-d4d7de7a7d44.png)

**Comment by boris-kolpackov** [comment]
> Furthermore, I believe it would be a good idea to provide a simple auto-formatting tool for build2 like clang-format for C++. In my opinion, such a tool maybe should be founded on the parser library of buildfiles.

If you mean reusing the `buildfile` parser that is part of the build system, then that won't be easy since it expands variables, wildcards, etc. On the other hard, the language doesn't have much formatting leeway (e.g., block `{` and `}` must be on separate lines) so maybe some ad hoc extension to the parser will be able to do the job.

**Comment by boris-kolpackov** [comment]
This issues came up again in the context of the Visual Studio extension being developed (https://github.com/build2/build2-vs) so I did a bit of research in this and related areas. Here is what I found:

Syntax highlighting is part of a broader problem of language support in IDEs/editors and includes auto-completion, go to definition, refactoring, etc., and the way to support these today seems to be via a Language Server Protocol (LSP)[1]. But, bizarrely, syntax highlighting is not part of the LSP specification, or, more precisely, handling of syntax highlighting by LSP is a very recent addition to the spec[2].

When it comes to syntax highlighting specifically, there seems to be two common approaches[3][4]: The old way is a TextMate regex-based grammar file (e.g., used by VSCode[5]). The new/future way is "semantic syntax highlighting" with the help of an LSP server (either via an extension to the protocol or hopefully in the near future via the new standard mechanism[2]). For example, it is my understanding that `rust-analyzer` currently does it via some ad hoc LSP extensions but will probably migrate to the stand way soon[4].

So it seems the future is in the LSP server but it's also a lot of work and I am not sure when we will get to it. Most likely not until we finish all the core features in the toolchain. At the same time, if anyone wants to help with that, I could probably find some time to cover the `build2`-specific tricky areas (like the "analyze" buildfile parsing mode and source fragment annotations).

The question is if we should try to come up with a TextMate grammar in the meantime (or for cases where LSP cannot be used for some reason -- I am still not clear how widespread TextMate grammar support is, for example, @kamrann will you be able to easily use it in a Visual Studio extension). I've also found some resources/examples on TextMate grammar[6][7]. On the surface it looks quite similar to @lyrahgames grammar mentioned above.












[1] https://en.wikipedia.org/wiki/Language_Server_Protocol
[2] https://github.com/microsoft/language-server-protocol/issues/18
[3] https://users.rust-lang.org/t/are-you-happy-with-the-current-state-of-rust-syntax-highlighting/49399
[4] https://github.com/rust-analyzer/rust-analyzer/issues/604
[5] https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide
[6] https://github.com/dustypomerleau/rust-syntax
[7] https://gist.github.com/JaimeStill/0497f283e09779da15c00be159175972

**Comment by boris-kolpackov** [comment]
> So if I'm understanding correctly, the idea with LSP would be that `b` (or some other executable linking against the core `build2` implementation) would implement the server side of the protocol, in C++ and reusing a lot of the machinery in place for the parsing of buildfiles that it already needs to do.

Yes, exactly. The build system is already a library with `b` being just a thin driver. We won't be able to re-use parsing as-is since it will have to be a special "analysis" mode rather than actual evaluation. But I think it still makes sense to do this as part of the normal parsing/evaluation code (we already have something like this in the form of the pre-parse mode). Current thinking is that you will be able to run the parser in this special analysis mode and get a sequence of semantically-annotated tokens with mapping to the source regions (e.g., something like "this token is a name which is a name of a variable in an expansion, it starts at position X and ends at position Y"). It looks like to implement LSP we may also need to have incremental analysis, which still requires thinking.

> In the shorter term, yes, TextMate grammar is apparently supported by Visual Studio out of the box.

Ok, I will try to look into this in the near future.

**Comment by kamrann** [comment]
> Current thinking is that you will be able to run the parser in this special analysis mode and get a sequence of semantically-annotated tokens with mapping to the source regions (e.g., something like "this token is a name which is a name of a variable in an expansion, it starts at position X and ends at position Y"). It looks like to implement LSP we may also need to have incremental analysis, which still requires thinking.

Sounds great. From what I can gather, in terms of supporting LSP in a Visual Studio extension, it's pretty much just a case of making a connection to the server; from there VS already implements the client side of the protocol itself and the associated editor integrations (for common functionality at least). So I guess we should at some point determine what constitutes a minimal server implementation of the protocol; from there we could get it integrated and then iteratively implement more functionality over time.

That said, I guess even a minimal implementation is probably a fair sized undertaking, so it probably makes sense to still start off with just a TextMate grammar.

**Comment by boris-kolpackov** [comment]
I could be wrong, but to me it seems that for the Buildfile language the most important functionality in this area would be the syntax highlighting. And as I understand, neither VSCode nor VS proper yet support doing this via the new LSP protocol additions and it's not clear how long it might take (doesn't sound like trivial functionality). So I think short-to-medium term TextMate grammar is probably the most realistic approach.

**Comment by alexchandel** [comment]
Releasing a traditional regex grammar now would go much further towards adoption and use than waiting for LSP later.  The more time that passes without tooling, the more would-be adopters lock themselves into something else.

That there is no syntax highlighting in the most popular text editor will be taken as a sign the project is moribund.

**Comment by boris-kolpackov** [comment]
Update on LSP and syntax highlighting (also often referred to as "semantic highlighting"): 

The LSP protocol added syntax highlighting in the latest version: https://microsoft.github.io/language-server-protocol/specification#textDocument_semanticTokens

VS Code now supports semantic highlighting via language servers: https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide

**Comment by Wmbat** [comment]
I'm not sure if this is relevant, but for my editor of choice (neovim). A very popular method of getting syntax highlighting has become this [plugin](https://tree-sitter.github.io/tree-sitter/) (which is also what github uses to highlight code apparently). It seems to work rather differently as the options listed above because it actually requires generating a whole parser for a specific CFG. 

I could try my hand at making a treesitter parser for `buildfile`s

**Comment by RedCMD** [comment]
here's my VSCode extension to help with authoring TextMate grammars https://marketplace.visualstudio.com/items?itemName=RedCMD.tmlanguage-syntax-highlighter
